x/mongo/driver/operation/hello_test.go

```diff
@@ -0,0 +1,664 @@
+
+		want := formatJSON(&clientMetadata{
+			Application: &application{Name: "foo"},
+			Driver:      &driver{Name: driverName, Version: version.Driver},
+			OS:          &dist{Type: runtime.GOOS},
+			Platform:    runtime.Version(),
+		})
+
+		assertDocsEqual(t, got, want)
+	})
+
+	t.Run("omit the platform", func(t *testing.T) {
+		// Calculate the full length of a bsoncore.Document.
+		temp, err := encodeClientMetadata("foo", maxClientMetadataSize)
+		require.NoError(t, err, "error constructing template: %v", err)
+
+		// Calculate what the environment costs.
+		edst, err := appendClientEnv(nil, false, false)
+		require.NoError(t, err, "error constructing env template: %v", err)
+
+		// Calculate what the os.type costs.
+		odst := bsoncore.AppendStringElement(nil, "type", runtime.GOOS)
+
+		// Calculate what the platform costs
+		pdst := appendClientPlatform(nil)
+
+		// Calculate what the environment plus the os.type costs.
+		envAndOSTypeAndPlatform := len(edst) + len(odst) + len(pdst)
+
+		got, err := encodeClientMetadata("foo", len(temp)-envAndOSTypeAndPlatform)
+		assert.Nil(t, err, "error in encodeClientMetadata: %v", err)
+
+		want := formatJSON(&clientMetadata{
+			Application: &application{Name: "foo"},
+			Driver:      &driver{Name: driverName, Version: version.Driver},
+			OS:          &dist{Type: runtime.GOOS},
+		})
+
+		assertDocsEqual(t, got, want)
+	})
+
+	t.Run("0 max len", func(t *testing.T) {
+		got, err := encodeClientMetadata("foo", 0)
+		assert.Nil(t, err, "error in encodeClientMetadata: %v", err)
+		assert.Len(t, got, 0)
+	})
+}
+
+func TestParseFaasEnvName(t *testing.T) {
+	clearTestEnv(t)
+
+	tests := []struct {
+		name string
+		env  map[string]string
+		want string
+	}{
+		{
+			name: "no env",
+			want: "",
+		},
+		{
+			name: "one aws",
+			env: map[string]string{
+				envVarAWSExecutionEnv: "hello",
+			},
+			want: envNameAWSLambda,
+		},
+		{
+			name: "both aws options",
+			env: map[string]string{
+				envVarAWSExecutionEnv:     "hello",
+				envVarAWSLambdaRuntimeAPI: "hello",
+			},
+			want: envNameAWSLambda,
+		},
+		{
+			name: "multiple variables",
+			env: map[string]string{
+				envVarAWSExecutionEnv:        "hello",
+				envVarFunctionsWorkerRuntime: "hello",
+			},
+			want: "",
+		},
+		{
+			name: "vercel and aws lambda",
+			env: map[string]string{
+				envVarAWSExecutionEnv: "hello",
+				envVarVercel:          "hello",
+			},
+			want: envNameVercel,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			for key, value := range test.env {
+				t.Setenv(key, value)
+			}
+
+			got := getFaasEnvName()
+			if got != test.want {
+				t.Errorf("parseFaasEnvName(%s) = %s, want %s", test.name, got, test.want)
+			}
+		})
+	}
+}
+
+func BenchmarkClientMetadata(b *testing.B) {
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			_, err := encodeClientMetadata("foo", maxClientMetadataSize)
```

Comment:
Optional: This benchmark covers the best-case scenario where all fields fit in the size limit. Consider adding a benchmark for the worst-case scenario where all of the optional fields are omitted (e.g. if app name is >512 bytes).